### ISSUE TITLE
Provide equivalent of Agent to manage ClientHttp2Sessions across requests

### ISSUE THREAD

**Issue by pietermees** [szmarczak/http2-wrapper#6]
To get to a fully transparent solution, on par with `Agent` for h1, we should look at providing a `ClientHttp2Session` manager that takes on the following responsibilities:
1. Establishing new connections when no connection is currently established or when the maximum concurrent streams has been reached on an existing `ClientHttp2Session`
2. Performing ALPN negotiation so we can transparently support both h1 and h2, without knowing upfront what the remote host supports
3. Detecting failed or closed sessions, and re-establishing them when needed
4. Have some forms of policing, like maximum simultaneous outgoing connections, socket timeouts etc
5. It would be great if it could also expose some metrics and statistics: how many h1 vs h2 connections, idle vs in use connections, multiplexing rate etc

We should expose both an `Agent` and new H2-oriented interface, so this connection pool can transparently support both protocols. This is necessary when we're connecting to a remote host and determine that it does not support h2, then we like to reuse that existing TLS connection over h1. If we expose an `Agent` interface`, we can reuse the TLS connection.

Ideally, this component is also DNS-aware, meaning that it could do the following:
- If DNS resolution indicates the remote host has multiple IP addresses, re-attempt connecting to a different IP if one connection fails. This makes sure that remote outages are handled more gracefully.
- If multiple connections need to be established to a remote host, do so across multiple of the available IPs so the load gets distributed.

**Comment by szmarczak** [comment]
All these are possible to do, but first of all I'm gonna create an Agent for HTTP2 [only].

Merging them so we could support H1 and H2 simultaneously is very hard to do, but possible :)

After I'm done making H2 Agent I'll begin merging H1 and H2 APIs.

**Comment by pietermees** [comment]
@szmarczak I pulled together some code that we already had internally and put it into a helper library: https://github.com/zentrick/alpn-agent
As the readme outlines, this could be used as a hybrid https / h2 agent with ALPN negotiation and share TLS session cache
The main problem with splicing this into the current `got` extension mechanism is that you only know which http stack you want to commit to (https or h2) after the ALPN negotiation finishes. I guess this could be resolved with a small update to `got` that accepts a `Promise` as a return value from `request` rather than just a synchronous response?

**Comment by szmarczak** [comment]
> I pulled together some code that we already had internally and put it into a helper library: https://github.com/zentrick/alpn-agent.

Yup, that's it.

> The main problem with splicing this into the current got extension mechanism is that you only know which http stack you want to commit to (https or h2) after the ALPN negotiation finishes.

Piece of cake.

> Alternatively, http2-wrapper could act as a proxy back to a regular https request instance if it is determined that h2 is not supported?

Not a good idea. I prefer the first one.

There's only one problem: our H2 Agent won't be compatible with H1 Agent.

**Comment by pietermees** [comment]
It is compatible right?

What we won’t be able to do is to transparently upgrade: magically making
an https request into h2 after ALPN negotiation. The agent interface
doesn’t allow us to swap out the request or response classes.

**Comment by szmarczak** [comment]
> It is compatible right? You can use this agent anywhere you use a standard https agent?

I think you're right but I'm not sure.

> What we won’t be able to do is to transparently upgrade: magically making
an https request into h2 after ALPN negotiation. The agent interface
doesn’t allow us to swap out the request or response classes.

I'll take care of it :)

**Comment by pietermees** [comment]
@szmarczak just wanted to check in and see if there was any progress on this :)

I noticed the [Node issue](https://github.com/nodejs/node/issues/17746) also mentioned you'd be submitting this code there as a PR?

It would probably make sense to prove out the transparent upgrade path in `got` by making that modification to allow `got` extensions to answer `request` with a `Promise`?

**Comment by szmarczak** [comment]
> just wanted to check in and see if there was any progress on this :)

The issue itself? Not yet, sorry for the delay. I'll try to focus on this project more. The project? yeah :smile: 

> I noticed the Node issue also mentioned you'd be submitting this code there as a PR?

Yeah, but don't expect it's going to land soon. I'm still finishing this project. TODO:

1. Rework `HTTP2ClientRequest`.
2. Make tests more reliable and readable.
   (maybe first stable release here?)
3. ALPN negotiation.
4. Create Agents for HTTP2.

> It would probably make sense to prove out the transparent upgrade path in got by making that modification to allow got extensions to answer request with a Promise?

I don't think it's really necessary. Anyway, we'll see :)

**Comment by szmarczak** [comment]
@pietermees I've sketched the [ALPN negotiation](https://github.com/szmarczak/http2-wrapper#http2autourl-options-callback). Could you test it please? (make sure to be on master branch) Any ideas/comments are welcome :smile_cat: 

> It would probably make sense to prove out the transparent upgrade path in got by making that modification to allow got extensions to answer request with a Promise?

You're right. I didn't see this coming. :clap: :clap: I'll make a PR tomorrow :D

**Comment by szmarczak** [comment]
The DNS-related functionalities shouldn't be built-in. But, our agent class should make it possible to do. See the examples (using HTTP1 Agent):

> If multiple connections need to be established to a remote host, do so across multiple of the available IPs so the load gets distributed.

https://gist.github.com/szmarczak/d29efa57d085fd7acc4e8b157f39cb45

> If DNS resolution indicates the remote host has multiple IP addresses, re-attempt connecting to a different IP if one connection fails. This makes sure that remote outages are handled more gracefully.

https://gist.github.com/szmarczak/8dcf43d4c57ff952424642afe0ebada2

**Comment by szmarczak** [comment]
> Establishing new connections when no connection is currently established or when the maximum concurrent streams has been reached on an existing ClientHttp2Session
> Have some forms of policing, like maximum simultaneous outgoing connections, socket timeouts etc

Currently, that's not possible. It would require to expose [`session[kState].pendingStreams`](https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js#L952)

**Comment by pietermees** [comment]
I think we can consider the case where max concurrent sessions is reached as a stretch goal as soon as Node exposes it. We should still be able to provide something in the meantime :)

**Comment by szmarczak** [comment]
> We should still be able to provide something in the meantime :)

We need to know if there is *any* pending stream though. Wait... It is possible! Even getting the amount of concurrent streams for a session! [This](https://nodejs.org/api/http2.html#http2_event_stream) may be a workaround ^_^

**Comment by pietermees** [comment]
Ah, but couldn’t you gracefully destroy the session with the close function
and waiting for the callback?
https://nodejs.org/dist/latest-v11.x/docs/api/http2.html#http2_http2session_close_callback

**Comment by pietermees** [comment]
Cross-posting this here: https://github.com/sindresorhus/got/issues/167#issuecomment-449882897
This [demo](https://gist.github.com/pietermees/cb9d8cc69e4c6bf05499aa8e1e6426c5) shows how to use `@zentrick/alpn-agent` together with `http2-wrapper` and `got`.

**Comment by szmarczak** [comment]
I've scratched some basic Pool implementation. These options are going to be present:

 - timeout - after some time of inactivity if there's no space for a free session it will be closed; if someone makes a new session and it reaches the limit it will close that session too; if there's no space it will throw an error,
 - maxSessions,
 - maxFreeSessions,
 - peerMaxConcurrentStreams - note that this is an initial value and will be replaced when the server sends a SETTINGS frame,
 - retry - if true it will make a new attempt to connect if there's available at least one free session, otherwise it will throw an error

If we only knew how to get the pending streams count... That would let us know when to create a new session. I haven't done this yet, but I have an idea: let's patch session.request so we will go through every stream. When called, increase pending streams + 1 and decrease when the `close` event is emitted.

**Comment by pietermees** [comment]
@szmarczak any particular reason why you'd not want to continue iterating on the `alpn-agent` package? Right now it only supports a single session, but [the intent is to expand on that](https://github.com/zentrick/alpn-agent/issues/3). 

PS: `peerMaxConcurrentStreams` is already supported btw

**Comment by szmarczak** [comment]
> PS: peerMaxConcurrentStreams is already supported btw

IKR :)

> any particular reason why you'd not want to continue iterating on the alpn-agent package?

First of all, I want to create a Pool that will work just out-of-box, so you can use it along with the native `http2` module. It'll support options mentioned above. Then we can discuss on how to create an Agent that will satisfy both protocols :) Keep in mind that HTTP2 sockets should **not** be malformed. So there will be no `keepAlive` option etc. The idea of `alpn-agent` package is very good - I didn't say I don't want to use it, it's still on the table ;) But as you've pointed out, there's some work to do.

**Comment by pietermees** [comment]
> First of all, I want to create a Pool that will work just out-of-box, so you can use it along with the native http2 module. It'll support options mentioned above.
Given that there is no `Agent` (or similar) interface for h2, I think any approach would basically come down to `alpn-agent` with some different API flavoring. You can already use the existing `createH2Session` API it offers to integrate natively to the NodeJS `http2` module.

> Then we can discuss on how to create an Agent that will satisfy both protocols :)
That's what I'm trying to understand. Given that the current implementation already supports this, there doesn't seem to be a need for the intermediate step and related duplication of work?

> Keep in mind that HTTP2 sockets should not be malformed. So there will be no keepAlive option etc. 
I assume you're talking about `Connection: keep-alive` headers? I agree that those should not be used with h2, but in `alpn-agent`, when talking about keep-alive, I'm referring to TCP keep-alive, which is applicable to h2.

> The idea of alpn-agent package is very good - I didn't say I don't want to use it, it's still on the table ;) But as you've pointed out, there's some work to do.
Absolutely, I guess I'm mostly bringing it up because I'd rather have us look at those next steps together than working on the same thing in parallel paths ;)

**Comment by szmarczak** [comment]
> I think any approach would basically come down to alpn-agent with some different API flavoring.

Node lacks the support of H2 Pool. I think firstly we should do something that can be merged into Node. It shouldn't be ambiguous.

> That's what I'm trying to understand. Given that the current implementation already supports this, there doesn't seem to be a need for the intermediate step and related duplication of work?

Nope. Agent is meant for HTTP1. It has different mechanism (similar, but different). IMO it should look like:

```js
const manager = new AgentManager({
	maxRequests: 100,
	maxConnections: 100,
	maxFreeConnections: 10,
	timeout: 60000,
	retry: true
});

manager.getAgent(); // returns Agent
manager.getPool(); // returns H2 Pool
```

Of course we could extend it and override some functions, like just normal Agent. `alpn-agent` is a compatibility layer, isn't it? :)

> I'm referring to TCP keep-alive, which is applicable to h2.

[Nah.](https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js#L723) I mean it is, but you mustn't change it.

> Absolutely, I guess I'm mostly bringing it up because I'd rather have us look at those next steps together than working on the same thing in parallel paths ;)

I'm happy to help :smiley: 

Can you take a look at this please? https://gist.github.com/szmarczak/f42dda503588ed314f4465acfa78c5ee
 I've scratched this today. It still needs lots of work to do. I just want to know what do you think :)

**Comment by pietermees** [comment]
> Node lacks the support of H2 Pool. I think firstly we should do something that can be merged into Node. It shouldn't be ambiguous.

Understood. I don't disagree here, it would be great to get a generic `Pool` implementation merged into Node. And I agree the easiest path to doing so is to make it as simple as possible :)

> Nope. Agent is meant for HTTP1. It has different mechanism (similar, but different). IMO it should look like:

Ok, got it. Makes sense.

> Of course we could extend it and override some functions, like just normal Agent. alpn-agent is a compatibility layer, isn't it? :)

I think the main piece here is that you want to negotiate ALPN first, then use the resulting socket from either an `https.Agent` or `http2.Pool` instance. But if we get a pool into Node, we can provide that as a layer on top of those two concepts and splice it in by using `createConnection`.

> Nah. I mean it is, but you mustn't change it.

Is there a particular reason why you shouldn't change it? I didn't run into those proxy errors yet because I was providing my own sockets that are already pre-configured.

> I'm happy to help 😃

Awesome, me too :)

> Can you take a look at this please? https://gist.github.com/szmarczak/f42dda503588ed314f4465acfa78c5ee
> I've scratched this today. It still needs lots of work to do. I just want to know what do you think :)

Provided first thoughts on the gist!

**Comment by szmarczak** [comment]
> But if we get a pool into Node, we can provide that as a layer on top of those two concepts and splice it in by using createConnection.

Sounds great!

> Is there a particular reason why you shouldn't change it? I didn't run into those proxy errors yet because I was providing my own sockets that are already pre-configured.

That's bad behavior. The module was designed to achieve best performance. Interfering the socket can make it slow, return invalid data, or, probably worst scenario: keep-alive turned off - the socket will close. There's no need to change them :)

H1 is an exception. We can make a new section for that:

```js
const manager = new AgentManager({
	maxRequests: 100,
	maxConnections: 100,
	maxFreeConnections: 10,
	timeout: 60000,
	http1: {
		keepAlive: false
	}
});
```

**Comment by szmarczak** [comment]
Also replied on the gist :)

**Comment by pietermees** [comment]
@szmarczak I just realized we might have a bit of a conceptual problem with tracking the _active_ and _idle_ states of a session:
1. an _active_ session becomes _idle_ when its last active stream ends
2. an _idle_ session becomes _active_ when we expect (?) a stream is going to start

It would make sense that we trigger the _active_ state when the number of active streams on the session increases from 0 to 1. So the invariant can be that an _idle_ session has 0 active streams and an _active_ session has a strictly positive number of active streams.

With our envisioned implementation right now, that's not the case. We set a session as _active_ when you first ask for a session, but you might never fire a request on it and it might therefore never be marked as _idle_?

I wonder whether the correct interface would actually be to expose a function with [ClientHttp2Session#request](https://nodejs.org/dist/latest-v11.x/docs/api/http2.html#http2_clienthttp2session_request_headers_options) semantics, which internally handles the creation, management and destruction of the backing sessions. It can also handle queueing of requests etc. It would also prevent people from messing up the pool by getting PUSH streams on idle sessions etc.

But in essence, that boils down to a `fetch` implementation for Node?

**Comment by szmarczak** [comment]
@pietermees Yup - the `idle` mechanism needs to be rewritten.

> I wonder whether the correct interface would actually be to expose a function with ClientHttp2Session#request

It would solve the problem of making requests... But H2 sessions are much more powerful - they need to be exposed.

Let's maybe move to Gitter to not mess up this place? :D

**Comment by szmarczak** [comment]
*Sorry I've deleted all our comments but I needed to clean up this place.*

I've rewritten the H2 Agent. [Take a look at it](https://github.com/szmarczak/http2-wrapper/blob/master/source/session-manager/index.js) (there are still some things to do, eg. correct session names).
This is **not** a generic pool. It's for `http2-wrapper` only.

> Establishing new connections when no connection is currently established or when the maximum concurrent streams has been reached on an existing ClientHttp2Session

I'll make a separate issue for this.

> Performing ALPN negotiation so we can transparently support both h1 and h2, without knowing upfront what the remote host supports

:heavy_check_mark: 

> Detecting failed or closed sessions, and re-establishing them when needed

We could retry failed sessions... but it's quite useless. Imagine all seats for free sockets are taken. You need to close one and make another... The retry logic should stay in a higher level packages like Got.

> Have some forms of policing, like maximum simultaneous outgoing connections, socket timeouts etc

:heavy_check_mark: 

> It would be great if it could also expose some metrics and statistics: how many h1 vs h2 connections, idle vs in use connections, multiplexing rate etc

It won't be compatible at all with the H1 Agent. Making a compatibility layer is a hard topic. The design completely differs.
